File: backend/models/parcel.py
from datetime import datetime
from typing import Any, Dict, Optional

from pydantic import BaseModel, Field, field_validator, model_validator

from models.common import Address, DeliveryMode, GeoPin, ParcelStatus, clean_optional_text


def clean_required_text(value: str) -> str:
    cleaned = clean_optional_text(value)
    if not cleaned:
        raise ValueError("Champ requis")
    return cleaned


class Parcel(BaseModel):
    parcel_id: str
    tracking_code: str
    sender_user_id: str
    recipient_phone: str
    recipient_name: str
    delivery_mode: DeliveryMode
    origin_relay_id: Optional[str] = None
    destination_relay_id: Optional[str] = None
    delivery_address: Optional[Address] = None
    weight_kg: float = 0.5
    dimensions: Optional[Dict[str, float]] = None
    declared_value: Optional[float] = None
    description: Optional[str] = None
    parcel_photo_url: Optional[str] = None
    parcel_photo_path: Optional[str] = None
    parcel_photo_file_id: Optional[str] = None
    parcel_photo_filename: Optional[str] = None
    parcel_photo_storage: Optional[str] = None
    parcel_photo_content_type: Optional[str] = None
    parcel_photo_uploaded_at: Optional[datetime] = None
    is_express: bool = False
    who_pays: str = "sender"
    quoted_price: float
    paid_price: Optional[float] = None
    payment_status: str = "pending"
    payment_method: Optional[str] = None
    payment_ref: Optional[str] = None
    promo_id: Optional[str] = None
    pickup_code: str = ""
    delivery_code: str = ""
    return_code: Optional[str] = None
    status: ParcelStatus = ParcelStatus.CREATED
    assigned_driver_id: Optional[str] = None
    redirect_relay_id: Optional[str] = None
    transit_relay_id: Optional[str] = None
    rating: Optional[int] = None
    rating_comment: Optional[str] = None
    driver_tip: float = 0.0
    external_ref: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    expires_at: Optional[datetime] = None


class ParcelCreate(BaseModel):
    recipient_phone: str = Field(..., min_length=8, max_length=32)
    recipient_name: str = Field(..., min_length=2, max_length=120)
    delivery_mode: DeliveryMode
    origin_relay_id: Optional[str] = Field(default=None, max_length=80)
    destination_relay_id: Optional[str] = Field(default=None, max_length=80)
    delivery_address: Optional[Address] = None
    transit_relay_id: Optional[str] = Field(default=None, max_length=80)
    origin_location: Optional[Address] = None
    weight_kg: float = Field(default=0.5, gt=0, le=1000)
    dimensions: Optional[Dict[str, float]] = None
    declared_value: Optional[float] = Field(default=None, ge=0, le=1_000_000_000)
    description: Optional[str] = Field(default=None, max_length=1000)
    external_ref: Optional[str] = Field(default=None, max_length=120)
    is_express: bool = False
    who_pays: str = Field(default="sender", pattern="^(sender|recipient)$")
    promo_id: Optional[str] = Field(default=None, max_length=80)
    initiated_by: str = Field(default="sender", pattern="^(sender|recipient)$")
    sender_phone: Optional[str] = Field(default=None, max_length=32)
    pickup_voice_note: Optional[str] = None
    delivery_voice_note: Optional[str] = None

    @field_validator("recipient_phone", "recipient_name")
    @classmethod
    def normalize_required_text_fields(cls, value: str) -> str:
        return clean_required_text(value)

    @field_validator(
        "origin_relay_id",
        "destination_relay_id",
        "transit_relay_id",
        "description",
        "external_ref",
        "promo_id",
        "sender_phone",
    )
    @classmethod
    def normalize_optional_text_fields(cls, value: Optional[str]) -> Optional[str]:
        return clean_optional_text(value)

    @model_validator(mode="after")
    def validate_delivery_mode_requirements(self):
        has_origin_gps = bool(self.origin_location and self.origin_location.geopin)
        has_delivery_address = self.delivery_address is not None

        if self.delivery_mode == DeliveryMode.RELAY_TO_RELAY:
            if not self.origin_relay_id or not self.destination_relay_id:
                raise ValueError("relay_to_relay requires origin_relay_id and destination_relay_id")
        elif self.delivery_mode == DeliveryMode.RELAY_TO_HOME:
            if not self.origin_relay_id:
                raise ValueError("relay_to_home requires origin_relay_id")
            if not has_delivery_address:
                raise ValueError("relay_to_home requires delivery_address")
        elif self.delivery_mode == DeliveryMode.HOME_TO_RELAY:
            if not has_origin_gps:
                raise ValueError("home_to_relay requires origin_location.geopin")
            if not self.destination_relay_id:
                raise ValueError("home_to_relay requires destination_relay_id")
        elif self.delivery_mode == DeliveryMode.HOME_TO_HOME:
            if not has_origin_gps:
                raise ValueError("home_to_home requires origin_location.geopin")
            if not has_delivery_address:
                raise ValueError("home_to_home requires delivery_address")

        return self


class ParcelEvent(BaseModel):
    event_id: str
    parcel_id: str
    event_type: str
    from_status: Optional[ParcelStatus] = None
    to_status: Optional[ParcelStatus] = None
    actor_id: Optional[str] = None
    actor_role: Optional[str] = None
    location: Optional[GeoPin] = None
    notes: Optional[str] = Field(default=None, max_length=1000)
    metadata: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime

    @field_validator("notes")
    @classmethod
    def normalize_notes(cls, value: Optional[str]) -> Optional[str]:
        return clean_optional_text(value)


class ParcelQuote(BaseModel):
    delivery_mode: DeliveryMode
    origin_relay_id: Optional[str] = Field(default=None, max_length=80)
    destination_relay_id: Optional[str] = Field(default=None, max_length=80)
    origin_location: Optional[Address] = None
    delivery_address: Optional[Address] = None
    weight_kg: float = Field(default=0.5, gt=0, le=1000)
    declared_value: Optional[float] = Field(default=None, ge=0, le=1_000_000_000)
    is_express: bool = False
    who_pays: str = Field(default="sender", pattern="^(sender|recipient)$")
    promo_code: Optional[str] = Field(default=None, max_length=80)

    @field_validator("origin_relay_id", "destination_relay_id", "promo_code")
    @classmethod
    def normalize_text_fields(cls, value: Optional[str]) -> Optional[str]:
        return clean_optional_text(value)

    @model_validator(mode="after")
    def validate_quote_requirements(self):
        has_origin_gps = bool(self.origin_location and self.origin_location.geopin)
        has_delivery_address = self.delivery_address is not None

        if self.delivery_mode == DeliveryMode.RELAY_TO_RELAY:
            if not self.origin_relay_id or not self.destination_relay_id:
                raise ValueError("relay_to_relay requires origin_relay_id and destination_relay_id")
        elif self.delivery_mode == DeliveryMode.RELAY_TO_HOME:
            if not self.origin_relay_id:
                raise ValueError("relay_to_home requires origin_relay_id")
            if not has_delivery_address:
                raise ValueError("relay_to_home requires delivery_address")
        elif self.delivery_mode == DeliveryMode.HOME_TO_RELAY:
            if not has_origin_gps:
                raise ValueError("home_to_relay requires origin_location.geopin")
            if not self.destination_relay_id:
                raise ValueError("home_to_relay requires destination_relay_id")
        elif self.delivery_mode == DeliveryMode.HOME_TO_HOME:
            if not has_origin_gps:
                raise ValueError("home_to_home requires origin_location.geopin")
            if not has_delivery_address:
                raise ValueError("home_to_home requires delivery_address")

        return self


class QuoteResponse(BaseModel):
    price: Optional[float] = None
    currency: str = "XOF"
    breakdown: Dict[str, Any] = Field(default_factory=dict)
    original_price: Optional[float] = None
    discount_xof: float = 0.0
    promo_applied: Optional[Dict[str, Any]] = None


class FailDeliveryRequest(BaseModel):
    failure_reason: str = Field(..., min_length=2, max_length=80)
    notes: Optional[str] = Field(default=None, max_length=1000)

    @field_validator("failure_reason")
    @classmethod
    def normalize_failure_reason(cls, value: str) -> str:
        return clean_required_text(value)

    @field_validator("notes")
    @classmethod
    def normalize_notes(cls, value: Optional[str]) -> Optional[str]:
        return clean_optional_text(value)


class RedirectRelayRequest(BaseModel):
    redirect_relay_id: str = Field(..., min_length=3, max_length=80)
    notes: Optional[str] = Field(default=None, max_length=1000)

    @field_validator("redirect_relay_id")
    @classmethod
    def normalize_redirect_relay_id(cls, value: str) -> str:
        return clean_required_text(value)

    @field_validator("notes")
    @classmethod
    def normalize_notes(cls, value: Optional[str]) -> Optional[str]:
        return clean_optional_text(value)


class ParcelRatingRequest(BaseModel):
    rating: int = Field(..., ge=1, le=5)
    comment: Optional[str] = Field(default=None, max_length=1000)
    tip: float = Field(default=0.0, ge=0, le=1_000_000)

    @field_validator("comment")
    @classmethod
    def normalize_comment(cls, value: Optional[str]) -> Optional[str]:
        return clean_optional_text(value)


class LocationConfirmPayload(BaseModel):
    lat: float = Field(..., ge=-90, le=90)
    lng: float = Field(..., ge=-180, le=180)
    accuracy: Optional[float] = Field(None, ge=0, le=10000)
    voice_note: Optional[str] = None
    label: Optional[str] = Field(default=None, max_length=240)
    district: Optional[str] = Field(default=None, max_length=120)
    city: Optional[str] = Field(default=None, max_length=120)
    notes: Optional[str] = Field(default=None, max_length=500)

    @field_validator("label", "district", "city", "notes")
    @classmethod
    def normalize_text_fields(cls, value: Optional[str]) -> Optional[str]:
        return clean_optional_text(value)


class AddressChangePreviewRequest(BaseModel):
    lat: float = Field(..., ge=-90, le=90)
    lng: float = Field(..., ge=-180, le=180)
    accuracy: Optional[float] = Field(None, ge=0, le=10000)
    voice_note: Optional[str] = None
    label: Optional[str] = Field(default=None, max_length=240)
    district: Optional[str] = Field(default=None, max_length=120)
    city: Optional[str] = Field(default=None, max_length=120)
    notes: Optional[str] = Field(default=None, max_length=500)

    @field_validator("label", "district", "city", "notes")
    @classmethod
    def normalize_text_fields(cls, value: Optional[str]) -> Optional[str]:
        return clean_optional_text(value)


class AddressChangeApplyRequest(AddressChangePreviewRequest):
    accept_surcharge: bool = False

File: mobile/lib/core/models/parcel.dart
class ParcelEvent {
  const ParcelEvent({
    required this.id,
    required this.parcelId,
    required this.eventType,
    required this.status,
    required this.createdAt,
    this.note,
    this.actorId,
  });

  final String id;
  final String parcelId;
  final String eventType;
  final String status; // valeur de to_status
  final DateTime createdAt;
  final String? note;
  final String? actorId;

  factory ParcelEvent.fromJson(Map<String, dynamic> json) => ParcelEvent(
        id: json['event_id'] as String? ?? json['id'] as String? ?? '',
        parcelId: json['parcel_id'] as String? ?? '',
        eventType: json['event_type'] as String? ?? '',
        // to_status contient le nouveau statut
        status: (json['to_status'] ?? json['status'] ?? 'created').toString(),
        createdAt: DateTime.tryParse(json['created_at']?.toString() ?? '') ??
            DateTime.now(),
        note: json['notes']?.toString(),
        actorId: json['actor_id']?.toString(),
      );
}

class Parcel {
  const Parcel({
    required this.id,
    required this.trackingCode,
    required this.status,
    required this.deliveryMode,
    required this.senderId,
    required this.createdAt,
    this.originRelayId,
    this.destinationRelayId,
    this.senderName,
    this.recipientName,
    this.recipientPhone,
    this.destinationAddress,
    this.destinationLat,
    this.destinationLng,
    this.weightKg,
    this.declaredValue,
    this.hasInsurance = false,
    this.totalPrice,
    this.paymentStatus,
    this.externalRef,
    this.events = const [],
    this.assignedDriverId,
    this.initiatedBy = 'sender',
    this.deliveryConfirmed = false,
    this.pickupConfirmed = false,
    this.deliveryLocation,
    this.pickupLocation,
    this.deliveryCode,
    this.pinCode,
    this.pickupCode,
    this.deliveryLat,
    this.deliveryLng,
    this.rating,
    this.ratingComment,
    this.driverTip = 0.0,
    this.recipientConfirmUrl,
    this.isRecipientView,
    this.promoId,
    this.senderPhotoUrl,
    this.recipientPhotoUrl,
    this.driverPhotoUrl,
    this.driverName,
    this.driverPhone,
    this.driverLocation,
    this.etaText,
    this.distanceText,
    this.etaSeconds,
    this.encodedPolyline,
    this.paymentMethod,
    this.whoPays,
    this.paymentOverride = false,
    this.deliveryBlockedByPayment = false,
    this.pickupVoiceNote,
    this.deliveryVoiceNote,
    this.addressChangeSurchargeXof = 0.0,
    this.driverBonusXof = 0.0,
    this.expiresAt,
  });

  final String id;
  final String trackingCode;
  final String status;
  final String deliveryMode;
  final String senderId;
  final String? originRelayId;
  final String? destinationRelayId;
  final String? senderName;
  final String? recipientName;
  final String? recipientPhone;
  final String? destinationAddress;
  final double? destinationLat;
  final double? destinationLng;
  final double? weightKg;
  final double? declaredValue;
  final bool hasInsurance;
  final double? totalPrice;
  final String? paymentStatus;
  final String? externalRef;
  final List<ParcelEvent> events;
  final String? assignedDriverId;
  final DateTime createdAt;
  // Confirmation GPS bidirectionnelle
  final String initiatedBy;
  final bool deliveryConfirmed;
  final bool pickupConfirmed;
  final Map<String, dynamic>? deliveryLocation; // {lat, lng, accuracy}
  final Map<String, dynamic>? pickupLocation;
  final String?
      deliveryCode; // code que le destinataire donne au livreur (domicile)
  final String?
      pinCode; // code que le destinataire donne au relais (retrait relais)
  final String?
      pickupCode; // code que l'expéditeur donne au livreur (H2R / H2H)
  final double? deliveryLat;
  final double? deliveryLng;
  final int? rating;
  final String? ratingComment;
  final double driverTip;
  final String? recipientConfirmUrl;

  /// Calculé côté backend : true si le viewer est le destinataire
  final bool? isRecipientView;
  final String? promoId;
  final String? senderPhotoUrl;
  final String? recipientPhotoUrl;
  final String? driverPhotoUrl;
  final String? driverName;
  final String? driverPhone;
  final Map<String, dynamic>? driverLocation;
  final String? etaText;
  final String? distanceText;
  final int? etaSeconds;
  final String? encodedPolyline;
  final String? paymentMethod;
  final String? whoPays;
  final bool paymentOverride;
  final bool deliveryBlockedByPayment;
  final String? pickupVoiceNote;
  final String? deliveryVoiceNote;
  final double addressChangeSurchargeXof;
  final double driverBonusXof;
  final DateTime? expiresAt;

  factory Parcel.fromJson(Map<String, dynamic> json) {
    // delivery_address est un objet Address { label, city, geopin:{lat,lng} }
    final deliveryAddr = json['delivery_address'] as Map<String, dynamic>?;
    final geopin = deliveryAddr?['geopin'] as Map<String, dynamic>?;

    return Parcel(
      id: json['parcel_id'] as String? ?? json['id'] as String? ?? '',
      trackingCode: json['tracking_code'] as String? ?? '',
      status: (json['status'] is String
              ? json['status']
              : (json['status'] as Map?)?['value'] ?? 'created')
          .toString(),
      deliveryMode: (json['delivery_mode'] is String
              ? json['delivery_mode']
              : (json['delivery_mode'] as Map?)?['value'] ?? 'relay_to_relay')
          .toString(),
      senderId: json['sender_user_id']?.toString() ?? '',
      senderName: json['sender_name']?.toString(),
      originRelayId: json['origin_relay_id']?.toString(),
      destinationRelayId: json['destination_relay_id']?.toString(),
      recipientName: json['recipient_name']?.toString(),
      recipientPhone: json['recipient_phone']?.toString(),
      destinationAddress: deliveryAddr?['label']?.toString() ??
          deliveryAddr?['district']?.toString(),
      destinationLat: (geopin?['lat'] as num?)?.toDouble(),
      destinationLng: (geopin?['lng'] as num?)?.toDouble(),
      weightKg: (json['weight_kg'] as num?)?.toDouble(),
      declaredValue: (json['declared_value'] as num?)?.toDouble(),
      hasInsurance: json['is_insured'] as bool? ?? false,
      totalPrice: (json['quoted_price'] as num?)?.toDouble() ??
          (json['paid_price'] as num?)?.toDouble(),
      paymentStatus: json['payment_status']?.toString(),
      externalRef: json['external_ref']?.toString(),
      events: (json['events'] as List<dynamic>?)
              ?.map((e) => ParcelEvent.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      assignedDriverId: json['assigned_driver_id']?.toString(),
      createdAt: DateTime.tryParse(json['created_at']?.toString() ?? '') ??
          DateTime.now(),
      initiatedBy: json['initiated_by']?.toString() ?? 'sender',
      deliveryConfirmed: json['delivery_confirmed'] as bool? ?? false,
      pickupConfirmed: json['pickup_confirmed'] as bool? ?? false,
      deliveryLocation: json['delivery_location'] as Map<String, dynamic>?,
      pickupLocation: json['pickup_location'] as Map<String, dynamic>?,
      deliveryCode: json['delivery_code'] as String?,
      pinCode: json['relay_pin'] as String? ?? json['pin_code'] as String?,
      pickupCode: json['pickup_code'] as String?,
      deliveryLat:
          (json['delivery_address'] as Map<String, dynamic>?)?['geopin'] != null
              ? ((json['delivery_address']['geopin']['lat']) as num?)
                  ?.toDouble()
              : null,
      deliveryLng:
          (json['delivery_address'] as Map<String, dynamic>?)?['geopin'] != null
              ? ((json['delivery_address']['geopin']['lng']) as num?)
                  ?.toDouble()
              : null,
      rating: json['rating'] as int?,
      ratingComment: json['rating_comment'] as String?,
      driverTip: (json['driver_tip'] as num?)?.toDouble() ?? 0.0,
      recipientConfirmUrl: json['recipient_confirm_url'] as String?,
      isRecipientView: json['is_recipient'] as bool?,
      promoId: json['promo_id'] as String?,
      senderPhotoUrl: json['sender_photo_url'] as String?,
      recipientPhotoUrl: json['recipient_photo_url'] as String?,
      driverPhotoUrl: json['driver_photo_url'] as String?,
      driverName: json['driver_name'] as String?,
      driverPhone: json['driver_phone'] as String?,
      driverLocation: json['driver_location'] as Map<String, dynamic>?,
      etaText: json['eta_text'] as String?,
      distanceText: json['distance_text'] as String?,
      etaSeconds: (json['eta_seconds'] as num?)?.toInt(),
      encodedPolyline: json['encoded_polyline'] as String?,
      paymentMethod: json['payment_method'] as String?,
      whoPays: json['who_pays'] as String?,
      paymentOverride: json['payment_override'] as bool? ?? false,
      deliveryBlockedByPayment:
          json['delivery_blocked_by_payment'] as bool? ?? false,
      pickupVoiceNote: json['pickup_voice_note'] as String?,
      deliveryVoiceNote: json['delivery_voice_note'] as String?,
      addressChangeSurchargeXof:
          (json['address_change_surcharge_xof'] as num?)?.toDouble() ?? 0.0,
      driverBonusXof: (json['driver_bonus_xof'] as num?)?.toDouble() ?? 0.0,
      expiresAt: DateTime.tryParse(json['expires_at']?.toString() ?? ''),
    );
  }

  bool get canBeCancelled => status == 'created';
  bool get isDelivered => status == 'delivered';
  bool get isRelayToHome => deliveryMode.endsWith('_to_home');
  bool get isRelayPickup => deliveryMode.endsWith('_to_relay');
}

class QuoteResponse {
  const QuoteResponse({
    required this.price,
    required this.currency,
    this.breakdown = const {},
    this.originalPrice,
    this.discountXof = 0.0,
    this.promoApplied,
  });

  final double price;
  final String currency;
  final Map<String, dynamic> breakdown;
  final double? originalPrice;
  final double discountXof;
  final Map<String, dynamic>? promoApplied;

  factory QuoteResponse.fromJson(Map<String, dynamic> json) => QuoteResponse(
        // Backend retourne { price, currency, breakdown }
        price: (json['price'] as num).toDouble(),
        currency: json['currency'] as String? ?? 'XOF',
        breakdown: json['breakdown'] as Map<String, dynamic>? ?? {},
        originalPrice: (json['original_price'] as num?)?.toDouble(),
        discountXof: (json['discount_xof'] as num?)?.toDouble() ?? 0.0,
        promoApplied: json['promo_applied'] as Map<String, dynamic>?,
      );

  double get base => (breakdown['base'] as num?)?.toDouble() ?? price;
  double get weight => (breakdown['weight'] as num?)?.toDouble() ?? 0;
  double get insurance => (breakdown['insurance'] as num?)?.toDouble() ?? 0;
}
